File: platform/platform-core/contracts/bootstrap_contract.py
"""
Bootstrap Contract (Python)

Defines the interface for FastAPI server bootstrap.
Bootstrap runs ONCE at application start/shutdown.

Interface:
    setup(config: BootstrapConfig) -> FastAPI
    shutdown(app: FastAPI) -> None
"""

import os
import re
from dataclasses import dataclass, field
from typing import Any


@dataclass
class LoggerConfig:
    """Logger configuration."""
    level: str = 'info'
    json_format: bool = True


@dataclass
class BootstrapConfig:
    """Configuration passed to bootstrap.setup()."""
    port: int = 52000
    host: str = '0.0.0.0'
    logger: LoggerConfig = field(default_factory=LoggerConfig)
    core_plugins: list[str] = field(default_factory=list)
    core_lifecycles: list[str] = field(default_factory=list)
    loaders: list[str] = field(default_factory=list)


@dataclass
class LoaderReport:
    """Structured report from a loader's execution."""
    loader: str = ''
    discovered: int = 0
    validated: int = 0
    imported: int = 0
    registered: int = 0
    skipped: int = 0
    errors: list = field(default_factory=list)
    details: dict = field(default_factory=dict)


def validate_bootstrap_config(config: BootstrapConfig) -> tuple[bool, list[str]]:
    """Validate a bootstrap config.

    Returns:
        Tuple of (is_valid, error_messages)
    """
    errors = []

    if not isinstance(config.port, int) or config.port < 1 or config.port > 65535:
        errors.append('port must be an integer between 1 and 65535')

    if not isinstance(config.host, str):
        errors.append('host must be a string')

    valid_levels = ('fatal', 'error', 'warn', 'warning', 'info', 'debug', 'trace', 'silent')
    if config.logger.level not in valid_levels:
        errors.append(f'logger.level must be one of: {", ".join(valid_levels)}')

    for key in ('core_plugins', 'core_lifecycles', 'loaders'):
        value = getattr(config, key)
        if not isinstance(value, list):
            errors.append(f'{key} must be a list')
        elif not all(isinstance(item, str) for item in value):
            errors.append(f'{key} must contain only strings')

    return (len(errors) == 0, errors)


def create_loader_report(loader_name: str) -> LoaderReport:
    """Create an empty loader report."""
    return LoaderReport(loader=loader_name)


def sort_by_numeric_prefix(file_paths: list[str]) -> list[str]:
    """Sort file paths by numeric prefix in filename.

    Files without numeric prefix sort to 999.
    """
    def extract_prefix(p: str) -> tuple[int, str]:
        match = re.match(r'^(\d+)', os.path.basename(p))
        return (int(match.group(1)) if match else 999, p)

    return sorted(file_paths, key=extract_prefix)

File: platform/platform-core/contracts/bootstrap-contract.mjs
/**
 * Bootstrap Contract
 *
 * Defines the interface for server bootstrap modules (both Fastify and FastAPI).
 * Bootstrap runs ONCE at application start/shutdown.
 *
 * Interface:
 *   setup(config: BootstrapConfig): Promise<Server>
 *   shutdown(server: Server): Promise<void>
 */

import path from 'node:path';

/**
 * @typedef {Object} BootstrapConfig
 * @property {number} port - Server port (Fastify: 51000, FastAPI: 52000)
 * @property {string} host - Server host (default: '0.0.0.0')
 * @property {Object} logger - Logger configuration
 * @property {string} logger.level - Log level (default: 'info')
 * @property {boolean} logger.prettyPrint - Pretty print for dev
 * @property {string[]} corePlugins - Plugin module paths to load
 * @property {string[]} coreLifecycles - Core lifecycle hook paths
 * @property {string[]} loaders - Loader module paths to execute
 */

/**
 * @typedef {Object} LoaderReport
 * @property {string} loader - Loader name
 * @property {number} discovered - Count of items found during discovery
 * @property {number} validated - Count of items that passed validation
 * @property {number} registered - Count of items successfully registered
 * @property {number} skipped - Count of items skipped
 * @property {Array<{path: string, step: string, error: string}>} errors - Error details
 * @property {Object} details - Additional loader-specific information
 */

/**
 * Validates a bootstrap config object.
 * @param {Object} config
 * @returns {{ valid: boolean, errors: string[] }}
 */
export function validateBootstrapConfig(config) {
  const errors = [];

  if (!config || typeof config !== 'object') {
    return { valid: false, errors: ['Config must be an object'] };
  }

  if (typeof config.port !== 'number' || config.port < 1 || config.port > 65535) {
    errors.push('port must be a number between 1 and 65535');
  }

  if (config.host !== undefined && typeof config.host !== 'string') {
    errors.push('host must be a string');
  }

  if (config.logger !== undefined) {
    if (typeof config.logger !== 'object') {
      errors.push('logger must be an object');
    } else if (config.logger.level !== undefined) {
      const validLevels = ['fatal', 'error', 'warn', 'info', 'debug', 'trace', 'silent'];
      if (!validLevels.includes(config.logger.level)) {
        errors.push(`logger.level must be one of: ${validLevels.join(', ')}`);
      }
    }
  }

  for (const key of ['corePlugins', 'coreLifecycles', 'loaders']) {
    if (config[key] !== undefined) {
      if (!Array.isArray(config[key])) {
        errors.push(`${key} must be an array`);
      } else if (!config[key].every((item) => typeof item === 'string')) {
        errors.push(`${key} must contain only strings`);
      }
    }
  }

  return { valid: errors.length === 0, errors };
}

/**
 * Creates a default bootstrap config.
 * @param {Partial<BootstrapConfig>} overrides
 * @returns {BootstrapConfig}
 */
export function createDefaultConfig(overrides = {}) {
  return {
    port: 51000,
    host: '0.0.0.0',
    logger: {
      level: process.env.LOG_LEVEL || 'info',
      prettyPrint: process.env.NODE_ENV !== 'production',
    },
    corePlugins: [],
    coreLifecycles: [],
    loaders: [],
    ...overrides,
  };
}

/**
 * Creates an empty loader report structure.
 * @param {string} loaderName
 * @returns {LoaderReport}
 */
export function createLoaderReport(loaderName) {
  return {
    loader: loaderName,
    discovered: 0,
    validated: 0,
    imported: 0,
    registered: 0,
    skipped: 0,
    errors: [],
    details: {},
  };
}

/**
 * Sort file paths by numeric prefix extracted from filename.
 * "01_config.lifecycle.mjs" -> prefix 1
 * "100_decorators.lifecycle.mjs" -> prefix 100
 * Files without numeric prefix sort to 999.
 * @param {string[]} filePaths
 * @returns {string[]}
 */
export function sortByNumericPrefix(filePaths) {
  return [...filePaths].sort((a, b) => {
    const numA = parseInt(path.basename(a).match(/^(\d+)/)?.[1] || '999', 10);
    const numB = parseInt(path.basename(b).match(/^(\d+)/)?.[1] || '999', 10);
    return numA - numB || a.localeCompare(b);
  });
}
